picard/ui/__init__.py

```diff
@@ -212,27 +212,22 @@ def show_help(self):
 # With py3, QObjects are no longer hashable unless they have
 # an explicit __hash__ implemented.
 # See: http://python.6.x6.nabble.com/QTreeWidgetItem-is-not-hashable-in-Py3-td5212216.html
-class HashableTreeWidgetItem(QtWidgets.QTreeWidgetItem):
-
+class HashableItem:
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.id = uuid.uuid4()
+        self.__id = uuid.uuid4()
+        self.__hash = hash(self.__id)
 
     def __eq__(self, other):
-        return self.id == other.id
+        return self.__id == other.__id
 
     def __hash__(self):
-        return hash(str(self.id))
-
+        return self.__hash
 
-class HashableListWidgetItem(QtWidgets.QListWidgetItem):
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.id = uuid.uuid4()
+class HashableTreeWidgetItem(HashableItem, QtWidgets.QTreeWidgetItem):
+    pass
 
-    def __eq__(self, other):
-        return self.id == other.id
 
-    def __hash__(self):
-        return hash(str(self.id))
+class HashableListWidgetItem(HashableItem, QtWidgets.QListWidgetItem):
```

Comment:
I'd turn this around to `class HashableListWidgetItem(QtWidgets.QListWidgetItem, HashableItem)`.

The reason is that I would first list the primary type we try to inherit. `HashableListWidgetItem` first and foremost is supposed to be a `QListWidgetItem`. The `HashableItem` class acts more as a mixin that ads extra functionality.

Same for the `HashableTreeWidgetItem` above.

Reply:
I tried, but when doing so, I get an error:

```
Traceback (most recent call last):
  File "/home/zas/src/picard/./picard/ui/mainwindow/__init__.py", line 905, in show_options
    options_dialog = OptionsDialog.show_instance(page, self)
  File "/home/zas/src/picard/./picard/ui/__init__.py", line 156, in show_instance
    instance = cls.get_instance(*args, **kwargs)
  File "/home/zas/src/picard/./picard/ui/__init__.py", line 150, in get_instance
    cls._instance = cls(*args, **kwargs)
  File "/home/zas/src/picard/./picard/ui/options/dialog.py", line 230, in __init__
    self.add_pages(None, default_pagename, self.ui.pages_tree)
  File "/home/zas/src/picard/./picard/ui/options/dialog.py", line 164, in add_pages
    self.item_to_page[item] = page
TypeError: unhashable type: 'HashableTreeWidgetItem'
```

Reply:
Ok, the wonders of multiple inheritance :shrug:

Reply:
Yes, they never really work as we expect... ;)